### PR TITLE
Add table privileges: #1830

### PR DESCRIPTION
**Pull request by mberhault**
- add privileges to the table descriptor, inherited from the DB descriptor
  at creation time.
- modify grant/revoke/show grants statements to operate on a table
- split grant sql logic test into grant_database and grant_table

Table-level permissions are not yet enforced.

### CODE REVIEW

File: sql/grant.go [hunk 1]
```diff
@@ -36,34 +35,25 @@ import (
...
 		return nil, err
 	}
 
 	// Now update the descriptor.
 	// TODO(marc): do this inside a transaction. This will be needed
 	// when modifying multiple descriptors in the same op.
-	descKey := keys.MakeDescMetadataKey(dbDesc.ID)
-	if err := p.db.Put(descKey, dbDesc); err != nil {
+	descKey := keys.MakeDescMetadataKey(descriptor.GetID())
+	if err := p.db.Put(descKey, descriptor); err != nil {
```
Comment (petermattis) [inline sql/grant.go]: Nifty how this works for both tables and databases.

Comment (MBerhault) [inline sql/grant.go]: yeah. Although the descriptorProto interface is getting rather crowded. I'll may tweak Privilege to be an interface.

File: sql/testdata/grant_database [hunk 2]
```diff
@@ -30,6 +24,18 @@ a        readwrite READ,WRITE
...
 a        test-user READ,WRITE
 
+# Create table to inherit DB permissions.
+statement ok
+CREATE TABLE a.t (id INT PRIMARY KEY)
+
+query TTT
+SHOW GRANTS ON a.t
+----
+Table    User      Privileges
```
Comment (petermattis) [inline sql/testdata/grant_database]: FYI, you'll have to either remove this line or add the `colnames` option after rebasing on top of my update to the logic test framework.

**Comment by petermattis** [comment]
LGTM

File: structured/structured.proto [hunk 3]
```diff
@@ -69,6 +69,14 @@ message IndexDescriptor {
   repeated uint32 column_ids = 5 [(gogoproto.customname) = "ColumnIDs"];
 }
 
+// PrivilegeDescriptor represents the sets of privileges on a descriptor.
+message PrivilegeDescriptor {
```
Comment (tamird) [inline structured/structured.proto]: I'm not sure I like the name descriptor here. How about just Privileges?

File: sql/parser/grant.go [hunk 4]
```diff
@@ -91,14 +91,18 @@ var (
...
+	Databases NameList
+	Tables    QualifiedNames
 }
 
 func (tl TargetList) String() string {
-	return fmt.Sprintf("%s %s", tl.Type, tl.Targets)
+	if tl.Databases != nil {
+		return fmt.Sprintf("DATABASE %s", tl.Databases)
+	}
+	return fmt.Sprintf("%s", tl.Tables)
```
Comment (tamird) [inline sql/parser/grant.go]: "TABLE"?

**Comment by tamird** [comment]
LGTM

File: sql/parser/grant.go (hunk 4 above)
Comment (MBerhault) [inline sql/parser/grant.go]: nope. the grammar supports both "GRANT|REVOKE ON TABLE t" and "GRANT|REVOKE ON t". However, we do not persist which one is being used, so we can't regenerate it. See related comment in parser_test

File: structured/structured.proto (hunk 3 above)
Comment (MBerhault) [inline structured/structured.proto]: we already have a bunch of privileges in parser. additionally, this matches the naming (see IndexDescriptor and ColumnDescriptor)

File: sql/testdata/grant_database (hunk 2 above)
Comment (MBerhault) [inline sql/testdata/grant_database]: done

**Pull request merged**